diff --git a/asn1x509-1.0.js b/asn1x509-1.0.js
index e5d66d9..f08b596 100644
--- a/asn1x509-1.0.js
+++ b/asn1x509-1.0.js
@@ -1807,7 +1807,6 @@ KJUR.asn1.x509.OID = new function(params) {
         'SHA256withDSA':        '2.16.840.1.101.3.4.3.2',
 
         'rsaEncryption':        '1.2.840.113549.1.1.1',
-        'subjectKeyIdentifier': '2.5.29.14',
 
         'countryName':          '2.5.4.6',
         'organization':         '2.5.4.10',
